Sources/PropertyEditor.swift

```diff
@@ -0,0 +1,680 @@
+	///
+	/// - returns: The result of the action.
+	public func modify<Result>(_ action: (inout Value) throws -> Result) rethrows -> Result {
+		var value: Value {
+			get { return commited.value }
+			set { self.try(newValue) }
+		}
+
+		return try rootBox.lock {
+			return try action(&value)
+		}
+	}
+
+	internal func revalidate() {
+		rootBox.lock {
+			action(commited.value)
+		}
+	}
+}
+
+extension PropertyEditor: BindingTargetProtocol {
+	public func consume(_ value: Value) {
+		self.try(value)
+	}
+}
+
+extension PropertyEditor: BindingSourceProtocol {
+	public func observe(_ observer: Observer<Value, NoError>, during lifetime: Lifetime) -> Disposable? {
+		return commited.producer
+			.take(during: lifetime)
+			.start(observer)
+	}
+}
+
+public protocol _PropertyEditorProtocol: class {
+	associatedtype Value
+	associatedtype ValidationError: Swift.Error
+
+	var property: PropertyEditor<Value, ValidationError> { get }
+
+	init<T: _PropertyEditorProtocol>(
+		_ inner: T,
+		transform: @escaping (T.Value) -> Value,
+		errorTransform: @escaping (T.ValidationError) -> ValidationError,
+		_ body: @escaping (Value) -> Result<T.Value, ValidationError>
+	)
+
+	init<T: _PropertyEditorProtocol>(
+		_ inner: T,
+		transform: @escaping (T.Value) -> Value,
+		_ body: @escaping (Value) -> Result<T.Value, ValidationError>
+	) where T.ValidationError == NoError
+}
+
+extension PropertyEditor: _PropertyEditorProtocol {
+	public var property: PropertyEditor<Value, ValidationError> {
+		return self
+	}
+}
+
+public enum ValidationResult<Value, Error: Swift.Error> {
+	case success(Value)
+	case failure(Value, Error)
+
+	public var isFailure: Bool {
+		if case .failure = self {
+			return true
+		} else {
+			return false
+		}
+	}
+
+	public var error: Error? {
+		if case let .failure(_, error) = self {
+			return error
+		} else {
+			return nil
+		}
+	}
+
+	public var result: Result<Value, Error> {
+		switch self {
+		case let .success(value):
+			return .success(value)
+		case let .failure(_, error):
+			return .failure(error)
+		}
+	}
+
+	fileprivate init<InnerValue>(_ innerValue: InnerValue, transform: (InnerValue) -> Value, validator: (Value) -> Result<InnerValue, Error>) {
+		let value = transform(innerValue)
+
+		switch validator(value) {
+		case .success:
+			self = .success(value)
+		case let .failure(error):
+			self = .failure(value, error)
+		}
+	}
+
+	public func map<U, E: Swift.Error>(_ transform: (Value) -> U, _ errorTransform: (Error) -> E, validator: (U) -> Result<Value, E>) -> ValidationResult<U, E> {
+		switch self {
+		case let .success(value):
+			switch validator(transform(value)) {
+			case let .success(innerValue):
+				return .success(transform(innerValue))
+			case let .failure(error):
+				return .failure(transform(value), error)
+			}
+
+		case let .failure(value, error):
+			return .failure(transform(value), errorTransform(error))
+		}
+	}
+}
+
+// FIXME: Remove the type parameter that works around type checker weirdness.
```

Comment:
It refers to the redundant `<()>` type parameter in the inheritance clause below. Without it weird things happen.